File: packages/sdk/examples/langchain/langgraph.py
"""
LangGraph + Veto — ToolNode Wrapper Example (Python)

This example shows how to add Veto guardrails to a raw LangGraph graph
by wrapping the ToolNode with Veto validation.

Setup:
    pip install veto-sdk langgraph langchain-openai
"""

import asyncio
from langgraph.graph import MessagesState, StateGraph, START, END
from langgraph.prebuilt import ToolNode, tools_condition
from langchain_openai import ChatOpenAI
from langchain_core.tools import tool
from veto import Veto, VetoOptions
from veto.integrations.langchain import create_veto_tool_node


@tool
def web_search(query: str) -> str:
    """Search the web for information."""
    return f"Results for: {query}"


@tool
def delete_file(path: str) -> str:
    """Delete a file from disk."""
    return f"Deleted {path}"


async def main():
    veto = await Veto.init(VetoOptions(api_key="your-key"))
    tools = [web_search, delete_file]

    model = ChatOpenAI(model="gpt-4o").bind_tools(tools)

    async def call_model(state: MessagesState):
        response = await model.ainvoke(state["messages"])
        return {"messages": [response]}

    # Wrap ToolNode with Veto validation
    tool_node = ToolNode(tools)
    veto_tool_node = create_veto_tool_node(
        veto,
        tool_node,
        on_deny=lambda name, args, reason: print(f"[Veto] Blocked {name}: {reason}"),
    )

    graph = StateGraph(MessagesState)
    graph.add_node("agent", call_model)
    graph.add_node("tools", veto_tool_node)
    graph.add_edge(START, "agent")
    graph.add_conditional_edges("agent", tools_condition, ["tools", END])
    graph.add_edge("tools", "agent")
    compiled = graph.compile()

    result = await compiled.ainvoke(
        {"messages": [{"role": "user", "content": 'Search for "veto sdk"'}]}
    )
    print(result["messages"][-1].content)


if __name__ == "__main__":
    asyncio.run(main())

File: packages/sdk/examples/langchain/langgraph.ts
/**
 * LangGraph + Veto — ToolNode Wrapper Example (TypeScript)
 *
 * This example shows how to add Veto guardrails to a raw LangGraph graph
 * without using the createAgent middleware API.
 *
 * Setup:
 *   npm install veto-sdk @langchain/langgraph @langchain/openai @langchain/core
 */

// @ts-expect-error example file, imports may not resolve in SDK repo
import { StateGraph, MessagesAnnotation, START, END } from '@langchain/langgraph';
import { ToolNode, toolsCondition } from '@langchain/langgraph/prebuilt';
import { ChatOpenAI } from '@langchain/openai';
import { tool } from '@langchain/core/tools';
import { z } from 'zod';
import { Veto } from 'veto-sdk';
import { createVetoToolNode } from 'veto-sdk/integrations/langchain';

// Define tools
const searchTool = tool(
  ({ query }) => `Results for: ${query}`,
  {
    name: 'web_search',
    description: 'Search the web',
    schema: z.object({ query: z.string() }),
  },
);

const deleteTool = tool(
  ({ path }) => `Deleted ${path}`,
  {
    name: 'delete_file',
    description: 'Delete a file from disk',
    schema: z.object({ path: z.string() }),
  },
);

async function main() {
  const veto = await Veto.init();
  const model = new ChatOpenAI({ model: 'gpt-4o' }).bindTools([searchTool, deleteTool]);

  const callModel = async (state: typeof MessagesAnnotation.State) => {
    const response = await model.invoke(state.messages);
    return { messages: [response] };
  };

  // Wrap the ToolNode with Veto validation
  const toolNode = new ToolNode([searchTool, deleteTool]);
  const vetoToolNode = createVetoToolNode(veto, toolNode, {
    onDeny: (name, _args, reason) => console.log(`[Veto] Blocked ${name}: ${reason}`),
  });

  const graph = new StateGraph(MessagesAnnotation)
    .addNode('agent', callModel)
    .addNode('tools', vetoToolNode)
    .addEdge(START, 'agent')
    .addConditionalEdges('agent', toolsCondition, ['tools', END])
    .addEdge('tools', 'agent')
    .compile();

  const result = await graph.invoke({
    messages: [{ role: 'user', content: 'Search for "veto sdk" then delete /tmp/cache' }],
  });

  console.log(result.messages.at(-1)?.content);
}

main().catch(console.error);
